Automattic/wp-calypso
# Theme search: Hide upsell banner when upsell empty search card is displayed

This PR hides the upsell banner when the upsell card is displayed to avoid having two nudges. 

#### Implementation details
I tried to implement the logic in a way that the [`renderBanner`](https://github.com/Automattic/wp-calypso/compare/update/remove-themes-nudge-empty-search?expand=1#diff-5e6c175ac60d9ed48935454e0bcf1861c5d5893c33a08edb946162d2e0dea550R238) method would know if it must be displayed or not based on a similar logic used in the `Empty` component [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/components/themes-list/index.jsx#L30) and [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/components/themes-list/index.jsx#L143) but that required that the themes search result state to be lifted up and that state is based on [complex selectors](https://github.com/Automattic/wp-calypso/blob/3470f222c94913a577aa9ba842d1bfa780952506/client/my-sites/themes/themes-selection.jsx#L210-L234), and triggered by the page load and [this component](https://github.com/Automattic/wp-calypso/blob/3470f222c94913a577aa9ba842d1bfa780952506/client/my-sites/themes/themes-selection.jsx#L159). That required a lot of refactoring and some regression impact to be thoroughly tested. For this use case, re-architecting this flow seems unnecessary.

Also, I tried to use [existing logic](https://github.com/Automattic/wp-calypso/blob/02e0301ad9348b638102742674af6e80e5250a91/client/state/themes/reducer.js#L385-L386) when populating the `themes` state to identify if the themes obtained in the search were empty. But due to caching, that logic did not update when the search query was cached, so the selector option tried above was the alternative.

Given the above, raising an event after the card is displayed and reading this piece of state in order to display or not the banner was a much simpler option that did the job.

#### Proposed Changes

* Add a piece of state that indicates if the Upsell Card is displayed
  * The piece of state includes the standard Redux components, action creator, reducer and selector.
* Fire the action to update the state when the Upsell card is displayed
* Read that piece of state and display the Upsell banner based on that (as well as existing logic).


#### Testing Instructions



* Navigate to `/themes` and select a _free_ site, URL should be `/themes/<free_site>`
* Make sure the Upsell banner is displayed 
![CleanShot 2022-10-18 at 08 43 06@2x](https://user-images.githubusercontent.com/3519124/196355263-5a399856-d5f6-4a8f-9b35-39c504672dc8.png)
* Search for a theme using term that does not return any result, for example, `noresult` or `Ocean`
* Make sure the Upsell banner is **not** displayed and the Upsell card is displayed
![CleanShot 2022-10-18 at 08 45 06@2x](https://user-images.githubusercontent.com/3519124/196355675-91da85e7-593a-4c31-98d9-ed0aefa0e776.png)
* Repeat the search using a term that returns results, for example, `Disco`
* Make sure the Upsell banner is displayed
* Switch to a non-free site, and navigate to themes, for example `/themes/<premium_site>`
* Make sure the Upsell banner is not displayed when searching for themes that return results and searching for themes that do not return results. Make sure the behaviour is the same than in Production

#### Pre-merge Checklist






Resolves #68789
